app/services/price_policies/time_based_price_calculator.rb

```diff
@@ -40,6 +45,47 @@ def cost_and_subsidy(duration_mins, discount_multiplier)
       end
     end
 
+    def duration_pricing_cost_and_subsidy(duration_mins)
+      duration_in_hours = duration_mins / 60.0
+
+      result = build_intervals.reduce({ time_left: duration_in_hours, cost: 0, subsidy: 0 }) do |acc, interval_data|
+        time_left = acc[:time_left]
+
+        interval_length = interval_data[:interval_end] - interval_data[:interval_start]
+
+        time_to_charge = [time_left, interval_length].min
+
+        acc[:time_left] -= time_to_charge
+        acc[:cost] += interval_data[:rate] ? interval_data[:rate] * time_to_charge : 0
+        acc[:subsidy] += interval_data[:subsidy] ? interval_data[:subsidy] * time_to_charge : 0
+
+        acc
+      end
+
+      { cost: result[:cost], subsidy: result[:subsidy] }
+    end
+
+    def build_intervals
+      sorted_duration_rates = price_policy.duration_rates.sort_by { |dr| dr.min_duration_hours || 1_000 }
+      lower_intervals = sorted_duration_rates.dup
+      higher_intervals = sorted_duration_rates.dup
+
+      if sorted_duration_rates.first.min_duration_hours > 0
+        lower_intervals.prepend nil
+      else
+        higher_intervals.shift
+      end
+
+      lower_intervals.zip(higher_intervals).map do |lower, higher|
+        hourly_rate = lower ? lower.rate : usage_rate * 60
+        hourly_subsidy = lower ? lower.subsidy : usage_subsidy * 60
+        interval_start = lower&.min_duration_hours || 0
+        interval_end = higher&.min_duration_hours || Float::INFINITY
+
```

Comment:
I'm wondering if the approach of building lower and higher intervals and zipping them is still worth the complexity.  Instead, I think we could do `sorted_duration_rates.each_with_index` and use the index to handle logic - like when index is 0, use default values for hourly rate, hourly subsidy, and interval start.  For interval end, we need to get min_duration_hours from the record at `index +1`.  This seems easier to reason about to me, what do you think?